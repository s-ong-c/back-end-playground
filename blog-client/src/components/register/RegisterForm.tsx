import * as React from 'react';
import styled from 'styled-components';
import LabelInput from '../common/LabelInput';
import useInputs from '../../lib/hooks/useInputs';
import RoundButton from '../common/RoundButton';

const RegisterFormBlock = styled.div`
    margin-top: 3rem;
    .buttons {
        margin-top: 6rem;
    }
`;

export type RegisterFormType = {
    displayName: string;
    email: string;
    username: string;
    shortBio: string;
}
export interface RegisterFormProps {
    onSubmit: (form: RegisterFormType) => any;
    defaultEmail: string | null | undefined;
}

const RegisterForm: React.SFC<RegisterFormProps> = ({onSubmit, defaultEmail}) => {
    const [form, onChange] = useInputs<RegisterFormType>({
        displayName: '', 
        email: 'somony9292@gmail.com', 
        username: '', 
        shortBio: ''
    })
    return (
        <RegisterFormBlock>
            <LabelInput 
                name="name"
                label="이름" 
                onChange={onChange}
                placeholder="이름을 입력하세요"
                value={form.displayName} 
                size={20}
            /> 
            <LabelInput 
                name="email"
                label="이메일" 
                onChange={onChange}
                placeholder="이메일을 입력하세요"
                value={form.email} 
                disabled={!!defaultEmail}
                size={25}
            />
            <LabelInput 
                name="username"
                label="아이디" 
                onChange={onChange}
                placeholder="아이디를 입력하세요"
                value={form.username} 
                size={15}
            /> 
            <LabelInput 
                name="shortBio"
                label="한 줄 소개" 
                onChange={onChange}
                placeholder="당신을 한 줄로 소개해보세요"
                value={form.shortBio} 
                size={25}
            /> 
            <div className="buttons">
                <RoundButton inline color="gray" to="/">취소</RoundButton>
                <RoundButton
                    inline
                    type="submit"
                    onClick={() =>
                        onSubmit({ ...form, email: defaultEmail || form.email })
                    }
                >
                    다음
                </RoundButton>
            </div>
        </RegisterFormBlock>
    );
};

export default RegisterForm;