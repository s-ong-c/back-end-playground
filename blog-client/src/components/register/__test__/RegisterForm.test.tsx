import * as React from 'react';
import { render, fireEvent } from 'react-testing-library';
import RegisterForm, { RegisterFormProps } from '../RegisterForm';
import { StaticRouter, MemoryRouter } from 'react-router';

describe('RegisterForm', () => {
  const setup = (props: Partial<RegisterFormProps> = {}) => {
    const initialProps: RegisterFormProps = {
      onSubmit: () => {},
      defaultEmail: null,
      error: null,
    };
    const utils = render(
      <MemoryRouter>
        <RegisterForm {...initialProps} {...props} />
      </MemoryRouter>,
    );
    const form = {
      displayName: utils.getByPlaceholderText(/이름을/) as HTMLInputElement,
      email: utils.getByPlaceholderText(/이메일을/) as HTMLInputElement,
      username: utils.getByPlaceholderText(/아이디를/) as HTMLInputElement,
      shortBio: utils.getByPlaceholderText(/당신을/) as HTMLInputElement,
    };
    const changeInputs = () => {
      fireEvent.change(form.displayName, {
        target: {
          value: '송민석',
        },
      });
      fireEvent.change(form.email, {
        target: {
          value: 'somony9292@gmail.com',
        },
      });
      fireEvent.change(form.username, {
        target: {
          value: 'songc',
        },
      });
      fireEvent.change(form.shortBio, {
        target: {
          value: '안녕하세요',
        },
      });
    };
    return {
      ...utils,
      form,
      changeInputs,
    };
  };
  it('renders property', () => {
    setup();
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('changes input', () => {
    const utils = setup();
    const { form, getByDisplayValue, changeInputs } = utils;
    changeInputs();
    getByDisplayValue('송민석');
    getByDisplayValue('somony9292@gmail.com');
    getByDisplayValue('songc');
    getByDisplayValue('안녕하세요');
  });
  it('submits form', () => {
    const onSubmit = jest.fn();
    const utils = setup({
      onSubmit,
    });
    const { getByText } = utils;
    utils.changeInputs();
    const button = getByText('다음');
    fireEvent.click(button);
    // expect(onSubmit).toBeCalledWith({
    //   displayName: '송민석',
    //   email: 'somony9292@gmail.com',
    //   username: 'songc',
    //   shortBio: '안녕하세요',
    // });
  });
  it('should show default email and lock', () => {
    const utils = setup({
      defaultEmail: 'somony9292@gmail.com',
    });
    expect(utils.form.email.value).toBe('somony9292@gmail.com');
    expect(utils.form.email.disabled).toBeTruthy();
    expect(utils.container.querySelector('svg')).toBeVisible();
  });
});
