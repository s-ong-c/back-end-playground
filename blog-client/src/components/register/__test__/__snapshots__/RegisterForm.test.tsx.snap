// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm matches snapshot 1`] = `
<div>
  <div
    class="sc-htpNat jeRpxP"
  >
    <div
      class="sc-bdVaJa eQDxPg"
    >
      <label>
        이름
      </label>
      <div
        class="group"
      >
        <div
          class="input-wrapper"
        >
          <input
            name="displayName"
            placeholder="이름을 입력하세요"
            size="20"
            value=""
          />
        </div>
        <div
          class="width-maker"
        >
          이름을 입력하세요
        </div>
      </div>
    </div>
    <div
      class="sc-bdVaJa eQDxPg"
    >
      <label>
        이메일
      </label>
      <div
        class="group"
      >
        <div
          class="input-wrapper"
        >
          <input
            name="email"
            placeholder="이메일을 입력하세요"
            size="25"
            value=""
          />
        </div>
        <div
          class="width-maker"
        >
          이메일을 입력하세요
        </div>
      </div>
    </div>
    <div
      class="sc-bdVaJa eQDxPg"
    >
      <label>
        아이디
      </label>
      <div
        class="group"
      >
        <div
          class="input-wrapper"
        >
          <input
            name="username"
            placeholder="아이디를 입력하세요"
            size="15"
            value=""
          />
        </div>
        <div
          class="width-maker"
        >
          아이디를 입력하세요
        </div>
      </div>
    </div>
    <div
      class="sc-bdVaJa eQDxPg"
    >
      <label>
        한 줄 소개
      </label>
      <div
        class="group"
      >
        <div
          class="input-wrapper"
        >
          <input
            name="shortBio"
            placeholder="당신을 한 줄로 소개해보세요"
            size="25"
            value=""
          />
        </div>
        <div
          class="width-maker"
        >
          당신을 한 줄로 소개해보세요
        </div>
      </div>
    </div>
    <div
      class="form-bottom"
    >
      <div
        class="buttons"
      >
        <button
          class="sc-bwzfXH kLyosE"
          color="gray"
        >
          취소
        </button>
        <button
          class="sc-bwzfXH eqBGAJ"
          color="teal"
          type="submit"
        >
          다음
        </button>
      </div>
    </div>
  </div>
</div>
`;
